### DISCUSSION TITLE
One to One relationship: Child primary key with parent foreign key

### DISCUSSION THREAD

**Discussion by RamonGiovane** [sqlalchemy/sqlalchemy#8363] [Usage Questions] [answered]
I'm  trying to do an one to one relationship in a way that both tables share the same id.
```python
class Parent:
    __tablename__ = 'parent'
     id = Column(Integer, primary_key=True)
     parent_info = Column(String)

class Child:
    __tabename__ = 'child'
    id = Column(Integer, ForeignKey(Parent.id), primary_key=True)
    child_info = Column(String)
    parent = relationship(
        'Parent',
        uselist=False,
        backref=backref('child', uselist=False)
    )
```
Is it possible? I'm running into some bugs, SQLAlchemy is emitting inserts to the parent table that I didn't asked for.
When I try:
```python
session.add(Child(child_info='info'))
await session.commit()
```

I'm expecting an `IntegrityError`, because I'm violating a null constraint (primary key that is also a non nullable foreign key), but it raises nothing and worse, it creates a new entry in the parent table with a new auto incremented ID and uses it as primary/foreign key in the child table. Like this:

```
--- parent -----           --- child -----
id    parent_info            id      child_info
1      NULL                   1       'info'
```

If you try to insert a new row with a RDBMS, it blocks normally, telling that there is no corresponding key in the parent table, meaning that the constraint is fine.

It really seems an ORM behavior, as you can clearly see the insert statement to the parent before inserting the child, if setting `echo=True`.

**Comment by zzzeek** [comment, marked as answer]
> I'm trying to do an one to one relationship in a way that both tables share the same id.
> 
> ```python
> class Parent:
>     __tablename__ = 'parent'
>      id = Column(Integer, primary_key=True)
>      parent_info = Column(String)
> 
> class Child:
>     __tabename__ = 'child'
>     id = Column(Integer, ForeignKey(Parent.id), primary_key=True)
>     child_info = Column(String)
>     parent = relationship(
>         'Parent',
>         uselist=False,
>         backref=backref('child', uselist=False)
>     )
> ```
> 
> Is it possible?

yup, sure

> 
> ```python
> session.add(Child(child_info='info'))
> await session.commit()
> ```
> 
> I'm expecting an `IntegrityError`, because I'm violating a null constraint (primary key that is also a non nullable foreign key), but it raises nothing and worse, it creates a new entry in the parent table with a new auto incremented ID and uses it as primary/foreign key in the child table. Like this:
> 
> ```
> --- parent -----           --- child -----
> id    parent_info            id      child_info
> 1      NULL                   1       'info'
> ```

the only way that would occur would be if you also created a Parent object.    Otherwise, the flush won't even proceed, SQLAlchemy will raise a FlushError for an object that has no primary key given or any means of generating one.

demo

```py
import asyncio

from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import backref
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship

Base = declarative_base()


class Parent(Base):
    __tablename__ = 'parent'
    id = Column(Integer, primary_key=True)
    parent_info = Column(String)

class Child(Base):
    __tablename__ = 'child'
    id = Column(Integer, ForeignKey(Parent.id), primary_key=True)
    child_info = Column(String)
    parent = relationship(
        'Parent',
        uselist=False,
        backref=backref('child', uselist=False)
    )


async def async_main():
    """Main program function."""

    engine = create_async_engine(
        "sqlite+aiosqlite://",
        echo=True,
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine) as session:
        session.add(Child(child_info='info'))
        await session.commit()

asyncio.run(async_main())
```

output:

```
sqlalchemy.orm.exc.FlushError: Instance <Child at 0x7f9cff610e50> has a NULL identity key.  If this is an auto-generated value, check that the database table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush() is not occurring at an inappropriate time, such as within a load() event.
```

**Reply by RamonGiovane** [reply]
Thank you! I will try to find out what I'm doing wrong.
